lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/evaluator/VortexWorker.java

```diff
@@ -82,37 +86,67 @@ public void run() {
-              final VortexRequest vortexRequest = VortexAvroUtils.toVortexRequest(message);
-              switch (vortexRequest.getType()) {
-              case ExecuteTasklet:
-                final TaskletExecutionRequest taskletExecutionRequest = (TaskletExecutionRequest) vortexRequest;
-                try {
-                  // Command Executor: Execute the command
-                  final Serializable result = taskletExecutionRequest.execute();
-
-                  // Command Executor: Tasklet successfully returns result
-                  final WorkerReport report =
-                      new TaskletResultReport<>(taskletExecutionRequest.getTaskletId(), result);
-                  workerReports.addLast(VortexAvroUtils.toBytes(report));
-                } catch (Exception e) {
-                  // Command Executor: Tasklet throws an exception
-                  final WorkerReport report =
-                      new TaskletFailureReport(taskletExecutionRequest.getTaskletId(), e);
-                  workerReports.addLast(VortexAvroUtils.toBytes(report));
-                }
-
-                heartBeatTriggerManager.triggerHeartBeat();
-                break;
-              default:
-                throw new RuntimeException("Unknown Command");
+          // Command Executor: Deserialize the command
+          final VortexRequest vortexRequest = VortexAvroUtils.toVortexRequest(message);
+
+          switch (vortexRequest.getType()) {
+            case ExecuteTasklet:
+              final CountDownLatch latch = new CountDownLatch(1);
+
+              // Scheduler Thread: Pass the command to the worker thread pool to be executed
+              // Record future to support cancellation.
+              futures.put(
+                  vortexRequest.getTaskletId(),
+                  commandExecutor.submit(new Runnable() {
+                    @Override
+                    public void run() {
+                      final TaskletExecutionRequest taskletExecutionRequest = (TaskletExecutionRequest) vortexRequest;
+                      try {
+                        // Command Executor: Execute the command
+                        final Serializable result = taskletExecutionRequest.execute();
+                        final WorkerReport report =
+                            new TaskletResultReport<>(taskletExecutionRequest.getTaskletId(), result);
+                        workerReports.addLast(VortexAvroUtils.toBytes(report));
+                      } catch (final InterruptedException ex) {
+                        // Assumes that user's thread follows convention that cancelled Futures
+                        // should throw InterruptedException.
+                        final WorkerReport report = new TaskletCancelledReport(taskletExecutionRequest.getTaskletId());
+                        LOG.log(Level.WARNING, "Tasklet with ID " + vortexRequest.getTaskletId() +
+                            " has been cancelled.");
```

Comment:
Here could be also `LOG.log(Level.WARNING, "Tasklet with ID {0} has been cancelled.", vortexRequest.getTaskletId());`